### ISSUE TITLE
`Style/RedundantRegexpEscape` auto-corrects `\w\-\#` to `\w-\#` which raises a SyntaxError

### ISSUE THREAD

**Issue by Narnach** [rubocop/rubocop#11355]
Applying Rubocop v1.41.1's auto-correct for `Style/RedundantRegexpEscape` on `%r{[\w\-\#]}` causes a `SyntaxError` due to the corrected Regexp being invalid: `unmatched range specifier in char-class: /[\w-\#]/ (SyntaxError)`

I validated this on Ruby 2.7.6, 2.7.7 and 3.1.2.

Looking through the changelog points me to #11150 which seems to have made relevant changes.

--------

## Expected behavior

I expect Rubocop's auto-correct to not cause invalid syntax due to the resulting Regexp being invalid.

## Actual behavior

Upgrading Rubocop from v1.39.0 to 1.41.1 caused a `Style/RedundantRegexpEscape` violation to trigger for this line in [CI](https://github.com/Narnach/groupie/pull/67):

```ruby
lib/groupie/tokenizer.rb:52:28: C: [Correctable] Style/RedundantRegexpEscape: Redundant escape inside regexp literal
      @raw.gsub!(%r{http[\w\-\#:/_.?&=]+}) do |url|
                           ^^
```

Auto-correcting it applied these changes:

```diff
-      @raw.gsub!(%r{http[\w\-\#:/_.?&=]+}) do |url|
+      @raw.gsub!(%r{http[\w-\#:/_.?&=]+}) do |url|
```

Attempting to load the "corrected" code now triggers a SyntaxError:

```
SyntaxError:
  unmatched range specifier in char-class: /http[\w-\#:\/_.?&=]+/
```

I've reduced the simplest test case in IRB to: `/[\w\-\#]/`, which causes the syntax error when "corrected" to `/[\w-\#]/`.

(Speculation) I suspect Rubocop only sees the `w` from `\w` and interprets the `\-` as intended for a range in that context.
If it would consider `\w` as the actual word character instead of just a letter, it can differentiate a case where `\-` is legitimately escaped.

## Steps to reproduce the problem

I've created a test case for `spec/rubocop/cop/style/redundant_regexp_escape_spec.rb` that fails because it suggests a correction:

```ruby
    context 'with a character class containing an escaped hyphen after a word character' do
      it 'does not register an offense' do
        expect_no_offenses('foo = %r{[\w\-\#]}')
      end
    end
```

## RuboCop version

I've used both the v1.41.1 release and latest commit (`908eda0df35b28e6f60311f47680208336c7ddaf`) to validate it the issue.

```
1.41.1 (using Parser 3.1.3.0, rubocop-ast 1.24.1, running on ruby 3.1.2) [x86_64-darwin21]
  - rubocop-performance 1.15.2
  - rubocop-rake 0.6.0
  - rubocop-rspec 2.16.0
```

**Comment by koic** [comment]
This issue is a dup with #11180.
